includes/amp-helper-functions.php

```diff
@@ -299,6 +299,85 @@ function amp_is_canonical() {
  */
 function amp_add_frontend_actions() {
 	add_action( 'wp_head', 'amp_add_amphtml_link' );
+
+	if ( AMP_Options_Manager::get_option( Option::MOBILE_REDIRECT ) && amp_version_is_available() ) {
+		// Insert the mobile redirect script as early as possible.
+		add_action( 'wp_head', 'amp_add_mobile_redirect_script', ~PHP_INT_MAX );
+	}
+}
+
+/**
+ * Add mobile redirect script.
+ *
+ * @since 1.6
+ */
+function amp_add_mobile_redirect_script() {
+	?>
+	<script>
+		(function ( ampSlug ) {
+			var hasTouchScreen;
+
+			if ("maxTouchPoints" in navigator) {
+				hasTouchScreen = navigator.maxTouchPoints > 0;
+			} else if ("msMaxTouchPoints" in navigator) {
+				hasTouchScreen = navigator.msMaxTouchPoints > 0;
+			} else {
+				var mQ = window.matchMedia && matchMedia("(pointer:coarse)");
+				if (mQ && mQ.media === "(pointer:coarse)") {
+					hasTouchScreen = !!mQ.matches;
+				} else if ('orientation' in window) {
+					hasTouchScreen = true; // deprecated, but good fallback
+				} else {
```

Comment:
I'm wary of this logic, as won't it mean laptops with touch screens will get redirected to the AMP version? If the user agent isn't used, then the minimum viewport dimension should be considered. Another option at some point would be to use `NetworkInformation.effectiveType`. Nevertheless, if a mobile visitor does land on the non-AMP version then they'll have the overhead of having to download the entire non-AMP HTML page to then get redirected to the AMP version. This being the case, it would actually be nice if a site could opt-in to doing the redirection server-side when `wp_is_mobile()` is true. The issue is that we can't rely on this by default due to variability in the network stack (as I understand).

Reply:
> I'm wary of this logic, as won't it mean laptops with touch screens will get redirected to the AMP version? If the user agent isn't used, then the minimum viewport dimension should be considered.

Yes, the viewport dimensions would definitely be taken into consideration as well. I think it would be best to settle on using the max viewport width of the browser to determine whether it's a mobile device or not. Mobile screens are getting bigger by the day so we'd have to consider a sane value to compare against. 

>  This being the case, it would actually be nice if a site could opt-in to doing the redirection server-side when wp_is_mobile() is true.

Perhaps we could add a filter here before the JS script is added to let users have the option of redirecting themselves. It would be better to provide a guaranteed solution here than a variant one.

Reply:
Yeah, this is tricky as it involves adding a filter to allow injecting JS code. If we provide a filter, I'd be more inclined to allow them to filter the list of user agents string tokens (or rather regular expression patterns) that cause redirection. This could then be exported to JS for use client-side but then also used to filter `wp_is_mobile` in case they opt-in to server-side redirection instead. Speaking of which, we should really provide a mechanism to do the server-side redirection if possible because redirection client-side is extremely slow, even though it is much more reliable. For sites that have administrators which are equipped technically able to verify that caching layers won't break server-side redirection, then they should be able to easily opt-in to server-side redirection via a filter. All this leads me to think that we really should limit detection to use agent sniffing, even though this has its own downsides.